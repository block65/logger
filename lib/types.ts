import { AsyncLocalStorage } from 'async_hooks';
import type pino from 'pino';
import type { MixinFn, MixinFnWithData } from './mixins.js';

export type Falsy = false | undefined | null;

export enum LogLevelNumbers {
  Fatal = 60,
  Error = 50,
  Warn = 40,
  Info = 30,
  Debug = 20,
  Trace = 10,
  Silent = 0,
}

export interface LogDescriptor {
  level: LogLevelNumbers;
  time: number;

  msg?: string;
  name?: string | undefined;

  // maye not be available depending on platform/cli
  pid?: number;
  hostname?: string;

  [key: string]: unknown;
}

type RemoveIndex<T> = {
  [K in keyof T as {} extends Record<K, 1> ? never : K]: T[K];
};

// pino tacks on a nasty `Record<string, any>` to its logger type
// we try to remove it here
export type BaseLogger = RemoveIndex<pino.Logger>;

export interface Logger extends BaseLogger {
  als: AsyncLocalStorage<AlsContext>;
}

export interface AlsContext {
  id: string;
  context?: Record<string, unknown>;
}

export interface AlsContextOutput {
  _contextId: AlsContext['id'];
  _context?: AlsContext['context'];
}

export type ComputePlatform = 'gcp-cloudrun' | 'aws-lambda' | 'aws';

export interface CreateLoggerOptions
  extends Omit<pino.LoggerOptions, 'mixin' | 'prettyPrint' | 'prettifier'> {
  pretty?: boolean;
  traceCaller?: boolean;
  platform?: ComputePlatform;
  mixins?: (MixinFnWithData | MixinFn | Falsy)[];
  transports?: pino.TransportMultiOptions;
}

export type CreateLoggerOptionsWithoutTransports = Omit<
  CreateLoggerOptions,
  'pretty' | 'transports'
>;

export interface CreateCliLoggerOptions
  extends Omit<CreateLoggerOptionsWithoutTransports, 'platform'> {
  color?: boolean;
}
