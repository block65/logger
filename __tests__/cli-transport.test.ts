import {
  afterEach,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import { LogLevelNumbers } from '../lib/types.js';
import { generateTmpFilenameAndReader, writeLogsToStream } from './helpers.js';

describe('CLI', () => {
  const oldEnv = process.env;

  beforeEach(() => {
    // jest.resetModules();
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2009-02-13T23:31:30.000Z'));
  });

  afterEach(() => {
    process.env = { ...oldEnv };
    jest.useRealTimers();
  });

  test('transport', async () => {
    const { cliTransport } = await import('../lib/transports/cli.js');

    const [dest, getLogs] = await generateTmpFilenameAndReader();
    const transport = await cliTransport({ dest });

    await writeLogsToStream(
      transport,
      {
        level: LogLevelNumbers.Info,
        time: Date.now(),
        pid: 1,
        hostname: 'test',
        msg: 'kek',
        random: {
          bool: true,
          str: 'hello',
          number: 12,
          arr: [1, 2, 3],
        },
      },
      {
        level: LogLevelNumbers.Error,
        time: Date.now(),
        msg: 'Kaboom',
        stack: ['stuff', 'code'],
      },
    );
    await expect(getLogs()).resolves.toMatchSnapshot();
  });

  test('logger force no-color', async () => {
    process.env.NO_COLOR = 'true';
    const { cliTransport } = await import('../lib/transports/cli.js');

    const [dest, getLogs] = await generateTmpFilenameAndReader();
    const transport = await cliTransport({ dest });

    await writeLogsToStream(
      transport,
      {
        level: LogLevelNumbers.Info,
        time: Date.now(),
        pid: 1,
        hostname: 'test',
        msg: 'kek',
        random: {
          bool: true,
          str: 'hello',
          number: 12,
          arr: [1, 2, 3],
        },
      },
      {
        level: LogLevelNumbers.Error,
        time: Date.now(),
        msg: 'Kaboom',
        stack: ['stuff', 'code'],
      },
    );

    await expect(getLogs()).resolves.toMatchSnapshot();
  });

  test('logger force-color', async () => {
    const { cliTransport } = await import('../lib/transports/cli.js');

    const [dest, getLogs] = await generateTmpFilenameAndReader();

    const transport = await cliTransport({
      dest,
      color: true,
    });

    await writeLogsToStream(
      transport,
      {
        level: LogLevelNumbers.Info,
        time: Date.now(),
        pid: 1,
        hostname: 'test',
        msg: 'kek',
        random: {
          bool: true,
          str: 'hello',
          number: 12,
          arr: [1, 2, 3],
        },
      },
      {
        level: LogLevelNumbers.Error,
        time: Date.now(),
        msg: 'Kaboom',
        stack: ['stuff', 'code'],
      },
    );

    await expect(getLogs()).resolves.toMatchSnapshot();
  });
});
