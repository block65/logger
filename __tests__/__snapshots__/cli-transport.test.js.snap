// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI logger force no-color 1`] = `
"2009-02-13T23:31:30.000Z INFO    kek { random: { arr: [ 1, 2, 3 ], bool: true, number: 12, str: 'hello' } }
2009-02-13T23:31:30.000Z ERROR   Kaboom { stack: [ 'stuff', 'code' ] }"
`;

exports[`CLI logger force-color 1`] = `
"[90m2009-02-13T23:31:30.000Z[39m [34mINFO[39m  [1mkek[22m { random: { arr: [ [33m1[39m, [33m2[39m, [33m3[39m ], bool: [33mtrue[39m, number: [33m12[39m, str: [32m'hello'[39m } }
[90m2009-02-13T23:31:30.000Z[39m [31mERROR[39m  [1mKaboom[22m { stack: [ [32m'stuff'[39m, [32m'code'[39m ] }"
`;

exports[`CLI transport 1`] = `
"2009-02-13T23:31:30.000Z INFO    kek { random: { arr: [ 1, 2, 3 ], bool: true, number: 12, str: 'hello' } }
2009-02-13T23:31:30.000Z ERROR   Kaboom { stack: [ 'stuff', 'code' ] }"
`;
